pkg/registry/helm.go

```diff
@@ -0,0 +1,363 @@
+		}
+	}
+
+	refSpec := &app.LibraryRefSpec{
+		Name:     partAlias,
+		Registry: h.Name(),
+	}
+
+	return part, refSpec, nil
+
+}
+
+// Name is the registry name.
+func (h *Helm) Name() string {
+	return h.spec.Name
+}
+
+// Protocol is the registry protocol.
+func (h *Helm) Protocol() Protocol {
+	return ProtocolHelm
+}
+
+// URI is the registry URI.
+func (h *Helm) URI() string {
+	return h.spec.URI
+}
+
+// IsOverride is true if this registry is an override.
+func (h *Helm) IsOverride() bool {
+	return h.spec.IsOverride()
+}
+
+// CacheRoot returns the root for caching by combining the path with the registry
+// name.
+func (h *Helm) CacheRoot(name string, relPath string) (string, error) {
+	return filepath.Join(name, relPath), nil
+}
+
+var (
+	validHelmSchemes = map[string]bool{
+		"http":  true,
+		"https": true,
+	}
+)
+
+// normalizeHelmURI normalizes a Helm repository URI by returning the
+// full URL to repository's index.yaml file.
+func normalizeHelmURI(s string) (string, error) {
+	u, err := url.Parse(s)
+	if err != nil {
+		return "", errors.Wrapf(err, "parsing Helm URI %q", s)
+	}
+
+	if _, ok := validHelmSchemes[u.Scheme]; !ok {
+		return "", errors.Errorf("%q is an invalid scheme for Helm repository", u.Scheme)
+	}
+
+	if strings.HasSuffix(u.Path, "index.yaml") {
+		return s, nil
+	}
+
+	u.Path = path.Join(u.Path, "index.yaml")
+	return u.String(), nil
+}
+
+// HelmRepository is metadata describing the contents of a Helm repository.
+type HelmRepository struct {
+	Charts map[string][]HelmRepositoryChart `json:"entries,omitempty"`
+}
+
+// Latest returns the latest version of charts in a Helm repository.
+func (hr *HelmRepository) Latest() []HelmRepositoryChart {
+	var out []HelmRepositoryChart
+
+	for _, chart := range hr.Charts {
+		if len(chart) > 0 {
+			sort.Slice(chart, func(i, j int) bool {
+				v1, err := semver.Make(chart[i].Version)
+				if err != nil {
+					return false
+				}
+				v2, err := semver.Make(chart[j].Version)
+				if err != nil {
+					return false
+				}
+
+				return v1.Compare(v2) == 1
+			})
+
+			out = append(out, chart[0])
+		}
+	}
+
+	return out
+}
+
+// HelmRepositoryChart is metadata describing a Helm Chart in a repository.
+type HelmRepositoryChart struct {
+	Description string   `json:"description,omitempty"`
+	Name        string   `json:"name,omitempty"`
+	URLs        []string `json:"urls,omitempty"`
+	Version     string   `json:"version,omitempty"`
+}
+
+type httpGet func(string) (*http.Response, error)
+
+type helmRepositoryClient struct {
+	url     string
+	httpGet httpGet
+}
+
+func newHelmRepositoryClient(u string, hg httpGet) (*helmRepositoryClient, error) {
+	u, err := normalizeHelmURI(u)
+	if err != nil {
+		return nil, errors.Wrap(err, "normalizing Helm repository URL")
+	}
+
+	if hg == nil {
+		hg = http.Get
+	}
+
+	return &helmRepositoryClient{
+		url:     u,
+		httpGet: hg,
+	}, nil
+}
+
+func (hrc *helmRepositoryClient) Repository() (*HelmRepository, error) {
```

Comment:
Let's document exported methods.

Reply:
This method isn't exported, but it should have documentation.

Reply:
Fair enough, and agreed 😃